istioctl/cmd/analyze.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"istio.io/istio/galley/pkg/config/analysis/analyzers"
+	"istio.io/istio/galley/pkg/config/analysis/local"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/source/kube/client"
+	"istio.io/pkg/log"
+
+	"github.com/spf13/cobra"
+)
+
+// Analyze command
+func Analyze() *cobra.Command {
+	analysisCmd := &cobra.Command{
+		Use:   "analyze <file|globpattern>...",
+		Short: "Analyze Istio configuration and print validation messages",
```

Comment:
What's the difference between `istioctl validate` and `istioctl analyze` commands? Seems like we'd want to merge them given the effort around validation convergence.

Reply:
Right now, the difference is that `istioctl validate` is focused on single-object validation, and operates only on files. `istioctl analyze` does multi-object validation, and operates on files, live clusters, or a combination of both. 

For now, we're keeping the experimental functionality separate, but I agree that ultimately we should end up with one command for users to run that does both.

Reply:
I understand the implementation is different. Is there a strong reason to keep the UX separate? If not, what would be the criteria for merging them later?

Reply:
The short-term reason is that we are trying to get the experimental command into 1.3.1, and want to keep it this sandboxed in an experimental space. 

I want to write up a more detailed merge plan in a design doc, but I expect that the most basic folding of `analyze` into `validate` will be pretty straightforward.

Reply:
Can you create a tracking issue for 1.4 and add a note here so users aren't confused why there are two commands that serve the same purpose?

Reply:
Created https://github.com/istio/istio/issues/16777 and added a comment linking to it.